### PR TITLE
Feature/new meeting emails #359

### PR DESCRIPTION
**Pull request by va-dym-ko-val**
image
image
image
image

### CODE REVIEW

**Review by Pavliuk13 — APPROVED** [review]

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/MeetingService.cs [hunk 1]
```diff
@@ -16,12 +16,19 @@ public class MeetingService : BaseService, IMeetingService
         private readonly IZoomService _zoomService;
         private readonly IEmailSenderService _emailSender;
         private readonly IGoogleMeetService _googleMeetService;
-        public MeetingService(EasyMeetsCoreContext context, IMapper mapper, IUserService userService, IZoomService zoomService, IEmailSenderService emailSender, IGoogleMeetService googleMeetService) : base(context, mapper)
+        private readonly IGoogleOAuthService _googleOAuthService;
+        private readonly ICalendarsService _calendarsService;
+        private const string _applicationURL = "https://bsa-easymeets.westeurope.cloudapp.azure.com/";
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/MeetingService.cs]: move to appsettings

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/MeetingService.cs [hunk 2]
```diff
@@ -142,6 +149,20 @@ public async Task<SaveMeetingDto> CreateMeeting(SaveMeetingDto meetingDto)
...
+            foreach (var calendar in visibleCalendars)
+            {
+                var tokenResultDto = await _googleOAuthService.RefreshToken(calendar.Calendar.RefreshToken);
+
+                if (tokenResultDto != null)
+                {
+                    await _calendarsService.AddMeetingToCalendar(_mapper.Map<SaveMeetingDto>(meeting), tokenResultDto);
+                }
+            }
+
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/MeetingService.cs]: Move to a separate method

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/MeetingService.cs [hunk 3]
```diff
@@ -187,15 +210,23 @@ public async Task SendEmailsAsync(long meetingId, TemplateType type)
...
+            {
+                emailDto = _mapper.Map<EmailDto>(emailTemplate);
+            }
 
             foreach (var recipient in recipients)
             {
+                if (emailTemplate is null)
+                {
+                    emailDto = GenerateNewMeetingEmailTemplate(meeting, recipient);
+                }
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/MeetingService.cs]: Why do we need two these checks?

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/MeetingService.cs [hunk 4]
```diff
@@ -250,5 +281,20 @@ public async Task DeleteMeeting(long meetingId)
...
+                Subject = $"New Event: {meeting.Name} - {meeting.StartTime} - {meeting.LocationType}",
+                Body = $"Hi, {member}!\n\n" +
+                $"A new event has been scheduled.\n" +
+                $"Event Type: {meeting.LocationType}\n" +
+                $"Invitee: {meeting.Author.Name}\n" +
+                $"Invitee Email: {meeting.Author.Email}\n" +
+                $"Event Date/Time: {meeting.StartTime}\n\n" +
+                $"View event in Easymeets: {_applicationURL}{meeting.MeetingLink}"
+            };
+        }
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/MeetingService.cs]: Create a separate method for parameters mapping, so we can reuse this mapping for other templates

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/MeetingService.cs (hunk 3 above)
Comment (va-dym-ko-val) [reply backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/MeetingService.cs]: We use this method for new meetings with "+ New Meeting" and Availability Slot. If there are any template, it`s "+ New Meeting" and method must generate template.

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/MeetingService.cs [hunk 5]
```diff
@@ -176,29 +190,32 @@ public async Task SendEmailsAsync(long meetingId, TemplateType type)
...
             }
 
-            var emailDto = _mapper.Map<EmailDto>(emailTemplate);
+            var emailTemplate = meeting?
+                .AvailabilitySlot?
+                .EmailTemplates
+                .FirstOrDefault(template => template.TemplateType == type && template.IsSend);
 
-            foreach (var recipient in recipients)
+            if (meeting is not null)
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/MeetingService.cs]: Do we need this check? we check for null in the 199 line

**Review by tatianahutii — CHANGES_REQUESTED** [review]

**Review by tatianahutii — APPROVED** [review]

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/MeetingService.cs [hunk 6]
```diff
@@ -175,29 +189,32 @@ public async Task SendEmailsAsync(long meetingId, TemplateType type)
...
```
Comment (apofis96) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/MeetingService.cs]: Move this check right after query to db and use early exit.

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/MeetingService.cs [hunk 7]
```diff
@@ -249,5 +266,65 @@ public async Task DeleteMeeting(long meetingId)
...
+        {
+            var visibleCalendars = await _context.CalendarVisibleForTeams
+                .Include(g => g.Calendar)
+                .Where(x => x.TeamId == meeting.TeamId).ToListAsync();
+
+            foreach (var calendar in visibleCalendars)
+            {
+                var tokenResultDto = await _googleOAuthService.RefreshToken(calendar.Calendar.RefreshToken);
+
+                if (tokenResultDto != null)
```
Comment (apofis96) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/MeetingService.cs]: is not

Comment (apofis96) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/MeetingService.cs:323]: is not

**Review by apofis96 — CHANGES_REQUESTED** [review]

**Review by apofis96 — APPROVED** [review]

**Pull request merged**